nri-prelude/src/Internal/IO.hs

```diff
@@ -0,0 +1,26 @@
+{-# LANGUAGE BlockArguments #-}
+
+module Internal.IO
+  ( putText,
+    putTextLn,
+  )
+where
+
+import Control.Concurrent.MVar
+import System.IO.Unsafe (unsafePerformIO)
+import Text
+import qualified Prelude
+
+{-# NOINLINE stdoutLock #-}
+stdoutLock :: MVar ()
+stdoutLock = unsafePerformIO (newMVar ())
+
+putText :: Text -> Prelude.IO ()
+putText text = do
+  withMVar stdoutLock \_ ->
+    Prelude.putStr (Text.toList text)
+
+putTextLn :: Text -> Prelude.IO ()
```

Comment:
If I was looking at a call site of this function, I would assume from the name that it was exactly the same as `putStrLn` except with `Text` instead of `String`. But in fact, if you use it, you are implicitly opting into some thread synchronisation stuff, and you may not realise it.

I think that is something that should be explicitly called out, perhaps by using a word like "lock" or "atomic" or something in the name.

Then if we ever somehow have a concurrency bug with this, it will at least "look suspicious".

Reply:
Most languages have some sort of lock on file descriptors while printing strings. I don't know why Haskell decided not to do so (actually there's a lock on the fd, but not while accessing the buffer which ends in the same issue).
We could definitely include some comments to clarify that this function actually provides some guarantees on the output not being interleaved with concurrent calls. But I wouldn't make it explicit in the name, since that's what we all expect from those functions anyway.